File: webmvp/src/lib/sectionHeaders.ts
/** Section label lines: `{Verse 1}` or `Verse 1:` — square brackets are chords only. */

const SECTION_LABEL_RE =
  /^(verse|chorus|bridge|intro|outro|tag|pre-chorus|instrumental|hook|refrain|breakdown|interlude|section)(\s+\d+)?$/i;

export function isSectionHeaderLine(line: string): boolean {
  const trimmed = line.trim();
  if (/^\{.+\}$/.test(trimmed)) {
    return true;
  }

  const colonMatch = trimmed.match(/^(.*?):$/);
  if (!colonMatch) {
    return false;
  }

  return SECTION_LABEL_RE.test(colonMatch[1].trim());
}

export function parseSectionHeaderLabel(line: string): string | null {
  const trimmed = line.trim();
  const curlyMatch = trimmed.match(/^\{(.+)\}$/);
  if (curlyMatch) {
    return curlyMatch[1].trim();
  }

  const colonMatch = trimmed.match(/^(.*?):$/);
  if (colonMatch && SECTION_LABEL_RE.test(colonMatch[1].trim())) {
    return colonMatch[1].trim();
  }

  return null;
}

File: mobile/lib/domain/section_headers.dart
final RegExp _sectionLabelRe = RegExp(
  r'^(verse|chorus|bridge|intro|outro|tag|pre-chorus|instrumental|hook|refrain|breakdown|interlude|section)(\s+\d+)?$',
  caseSensitive: false,
);

bool isSectionHeaderLine(String line) {
  final trimmed = line.trim();
  if (RegExp(r'^\{.+\}$').hasMatch(trimmed)) {
    return true;
  }
  final colonMatch = RegExp(r'^(.*?):$').firstMatch(trimmed);
  if (colonMatch == null) {
    return false;
  }
  return _sectionLabelRe.hasMatch(colonMatch.group(1)!.trim());
}

String? parseSectionHeaderLabel(String line) {
  final trimmed = line.trim();
  final curlyMatch = RegExp(r'^\{(.+)\}$').firstMatch(trimmed);
  if (curlyMatch != null) {
    return curlyMatch.group(1)!.trim();
  }
  final colonMatch = RegExp(r'^(.*?):$').firstMatch(trimmed);
  if (colonMatch != null &&
      _sectionLabelRe.hasMatch(colonMatch.group(1)!.trim())) {
    return colonMatch.group(1)!.trim();
  }
  return null;
}
